### ISSUE TITLE
Add QuickUnlock Feature

### ISSUE THREAD

**Issue by seatedscribe** [keepassxreboot/keepassxc#488]
With the newest and coolest feature of passphrase generator finally merged in, I find this request even more opportune - as you could tell, I have a LONG passphrase as master key for any of my databases, so my opinion may be a little biased :smile:

A rationale for this feature request is best explained by the author of **keepass2android**. I use this app regularly on a daily base and could not even think of practically handling the unlock procedure without this service. Here it goes, step by step (courtesy of [https://keepass2android.codeplex.com/](url))

_You should protect your password database with a strong (i.e. random and LONG) password including upper and lower case as well as numbers and special characters. Typing such a password ~~on a mobile phone~~ every time you unlock your database is time-consuming and error-prone._

- _Use a strong password for your database_
- _Load your database and type the strong password once. Enable QuickUnlock_
- _The application is locked after the time specified in the settings_
- _If you want to re-open your database, you can type just a few characters (by default, the last 3 characters of your password) to unlock quickly and easily!_
- _If the wrong QuickUnlock key is entered, the database is locked and the full password is required to re-open_

_Is this safe? First: it allows you to use a really strong password, this increases safety in case someone gets your database file. Second: If ~~you loose your phone~~ someone seize your laptop and tries to open your password database, the attacker has exactly one chance to make use of QuickUnlock. When using 3 characters and assuming 70 characters in the set of possible characters, the attacker has a 0.0003% chance of opening the file. If this sounds still too much for you, choose 4 or more characters in the settings._

I am not an expert on how this should be done correctly from a crypto perspective, but I am available and open to collaborations so to have this feature implemented :white_check_mark:

**Comment by StefanB08** [comment]
I agree that it is practical and use it myself, but it do actually makes your DB a lot less safe against a advanced attacker. The safety example is only true if you do not take cloning/imaging into account. By cloning the unit, the attacker will have a indefinite tries to enter the correct password. And as you have practically shortened your 70 character keylength down to a very weak 3 character password it will be possible to unlock the unit in a very little amount of time.

By using the quick unlock there is also a chance that the user will lock themselves out of the DB if they forget the true password and is only used to enter the short version of the full password.

I'm not against quick unlock, but I strongly believe it's critical to not mislead the user into thinking they are more or even equally secure by using it. Because they are definitely not.

**Comment by TheZ3ro** [comment]
I think this is a good thing for an opt-in option with the ability to select the length.

Like @StefanB08 suggested, if your adversary have great capabilities this will effectively lower your security. For the average user I don't think it's a great problem so opt-in is fine by me.

Also a side note, this option will need the master password stored in memory (for decrypting the database when the QuickUnlock is entered)

**Comment by bddenhartog** [comment]
I'd be much more a fan of enabling QuickUnlock with some other identifier, such as a separately-set `PIN`. This PIN can be a configurable length (it should probably default to 4), and can only be used to unlock the database during the current KPXC process if the database has been unlocked before -- to ensure the `PIN` (the "QuickUnlock passphrase") cannot be used to unlock the database from a new process / session.

**Comment by seatedscribe** [comment]
@StefanB08 I agree with you that in this scenario (a laptop in the hands of the cruel enemy) quick unlock is the least feature one wants enabled for his password manager. But at the same time we can think of alternatives, like having also a yubikey inserted which can reduce the risk of a brute force attack. As before, this should be an opt-in depending on the paranoid level of the user :smile: 

@bddenhartog having a PIN unrelated to the master password reduces the attack surface. Should the PIN remain the same from session to session, or is it to be generated anew every time KXC is launched?
 - In the latter case it could be challenging (or a good memory exercise) since it changes frequently; 
 - In the former case, where is it supposed to be stored in the DB? @TheZ3ro maybe KDBX4 can be extended to allow this feature? Just sayin'...

**Comment by bddenhartog** [comment]
quote @seatedscribe:

>`@bddenhartog` having a PIN unrelated to the master password reduces the attack surface. Should the PIN remain the same from session to session, or is it to be generated anew every time KXC is launched?
>
> * In the latter case it could be challenging (or a good memory exercise) since it changes frequently;
> * In the former case, where is it supposed to be stored in the DB? `@TheZ3ro` maybe KDBX4 can be extended to allow this feature? Just sayin'...

It would definitely be _most convenient_ to store the (non-ephemeral) PIN in the database settings; and I agree that this would be the preferred method of storage (over say, storing it on disk [permanent] or in memory [ephemeral]).

As a workaround to the current impossibility of storing the `PIN` in the database settings, perhaps we could read the PIN from the CLI (not unlike `--pw-stdin` does), and/or look for the value in a `.keepassxc` configuration file in the user's home directory?

**Comment by TheZ3ro** [comment]
Implementation wise, all those options are not available to us.

We have 2 cases:
1) Using a PIN not related to the master password (stored somewhere, doesn't matter now)
2) Using some character from the master password

In both those cases, if the master password is not stored in memory, you won't be able to unlock the database. By having a PIN or some char of the master password you can't derive the complete master password. (The DB is directly encrypted with the masterpassword)
And also if your database is locked with a Yubikey, you can't unlock it in both cases above if you don't have the Yubikey inserted.

**Comment by droidmonkey** [comment]
@thez3ro i believe you use the pin or last chars to encrypt the composite key in memory. Am i correct?

**Comment by TheZ3ro** [comment]
@droidmonkey should be a viable option, but then you shouldn't store the PIN anywhere and decrypt on demand

**Comment by droidmonkey** [comment]
The pin is known by the end user thus not stored.

**Comment by TheZ3ro** [comment]
@droidmonkey yes, I was clarifying this:

> As a workaround to the current impossibility of storing the PIN in the database settings, perhaps we could read the PIN from the CLI (not unlike --pw-stdin does), and/or look for the value in a .keepassxc configuration file in the user's home directory?

**Comment by TheZ3ro** [comment]
Closed accidentally, my bad. Reopened

**Comment by ArchangeGabriel** [comment]
With regard to PIN QuickUnlock, I was wondering if we could somehow have TOTP QuickUnlock? That would be interesting for me. Getting a TOTP token from my phone would still be way faster than typing my full passphrase. I know since @TheZ3ro answer to my comment [here](https://github.com/keepassxreboot/keepassxc/pull/380#issuecomment-291740749) that TOTP is not suitable for initial unlocking, but maybe it could be here?

**Comment by Keisial** [comment]
QuickUnlock protects the master secret with the "easier" one (precisely so you don't have to type your full passphrase). You can't do that with TOTP for the reasons explained in that comment.

You may want to open a new request to have an option of not actually locking keepassxc but leaving everything in memory awaiting a TOTP token.

**Comment by Keisial** [comment]
@TheZ3ro What's the problem with storing the PIN in the database itself? It's probably less sensitive than other db contents. It could be stored with no view interface, so that brief access to an unlocked instance doesn't allow later compromises.

**Comment by TheZ3ro** [comment]
@Keisial we don't have such a filter, all entry in the DB are displayed. We can store it like *KeePassHTTP Settings* but then you will need the masterpassword to access it/read it.

A viable solution can be: 
Store a new entry called *KeePassXC QuickUnlock PIN* as the masterpassword encrypted with the user-selected PIN (I will call this QuickSecret). This will be stored in the DB encrypted with the masterpassword itself.
Then once the DB is unlocked for the first time the QuickSecret is loaded in memory. Whenever the DB is locked again KeePassXC will ask for the PIN, decrypt the QuickSecret to get the masterpassword and the unlock the DB. If the PIN is wrong, delete from memory the QuickSecret so an attacker has only 1 attempt.

The only problem in this setup is that what we call "masterpassword" can be: 1 password, 1 keyfile and/or 1 yubikey.

Sincerely, I don't really like adding another way to unlock my databases so I think we should get this right if users want this feature.
I will look into KeePass2Android code to see their implementation.

**Comment by dok18** [comment]
Any news about this?

**Comment by droidmonkey** [comment]
This is a wishlist item, no plans as of right now to implement.

**Comment by TheZ3ro** [comment]
I was looking into this and noticed that KeePass2Android does the QuickUnlock by asking X character from the master-password

The master-password is stored in memory in plain-text [1], the masterkey (master-password + keyfile + yubikey) is also stored in memory so it can unlock the Db just fine [2]

This implementation will not work with KeePassXC since once the user inserted his master-password, the plain-text is thrown away and only an hash of it is kept in memory.
So we can't really ask for the first 10 digit of the master-password since KeePassXC doesn't have this information and can't check if it's correct. Only the full master-password hash will match.

I will think about it


--- 
[1] https://github.com/PhilippC/keepass2android/blob/17cbb754a6d46cc3dfae3399804d796014b3cddf/src/KeePassLib2Android/Keys/KcpPassword.cs#L77
[2] https://github.com/PhilippC/keepass2android/blob/4a738990ae9807bb30629df9551650e99ee82640/src/keepass2android/QuickUnlock.cs#L323

**Comment by seatedscribe** [comment]
> once the user inserted his master-password, the plain-text is thrown away and only an hash of it is kept in memory.

What about storing the hash of the last n characters along with the hash of the full master-password? This way the quick password can be processed the same way as the full-size password...
Basically I'm saying the db is protected with a master password which is a subset (last n chars) of the original, but only one attempt is permitted. Is it still too weak of an approach?
(please be kind, I'm no expert here) :)

**Comment by mydexterid** [comment]
There's another implementation where you can choose between a part of your master password or an entry in the database itself:
https://github.com/JanisEst/KeePassQuickUnlock

**Comment by mydexterid** [comment]
@TheZ3ro 
> The master-password is stored in memory in plain-text [1], the masterkey (master-password + keyfile + yubikey) is also stored in memory so it can unlock the Db just fine [2]

I don't know how did you come to a conclusion of it is stored in plain-text.
The linked linked line shows:
`m_psPassword = new ProtectedString(true, pbPasswordUtf8);`
and if you go a bit further:
https://github.com/PhilippC/keepass2android/blob/17cbb754a6d46cc3dfae3399804d796014b3cddf/src/KeePassLib2Android/Security/ProtectedString.cs#L175
https://github.com/PhilippC/keepass2android/blob/17cbb754a6d46cc3dfae3399804d796014b3cddf/src/KeePassLib2Android/Security/ProtectedBinary.cs#L288

So ProtectedString with utf8 byte array seems pretty much encrypted to me.

**Comment by phoerious** [comment]
The string itself may be encrypted, but the decryption key for it has to be stored in plain text. Otherwise you wouldn't be able to access the data.

**Comment by mydexterid** [comment]
> The string itself may be encrypted, but the decryption key for it has to be stored in plain text. Otherwise you wouldn't be able to access the data.

I only talked about the master-password. Which _is_ encrypted (not may be). Though you're right about the encryption key (there are some other protections in the lib itself for Windows, but that does not matter on Android.)
Also the user provided quickunlockpin _can_ be used for the encryption key, like it is done here:
https://github.com/JanisEst/KeePassQuickUnlock/blob/master/QuickUnlockProvider.cs#L135

**Comment by phoerious** [comment]
There is just no secure in-memory encryption. In-memory encryption just means that you have a key somewhere in another memory location of which you hope that an attacker has no access to it. But without proper hardware and OS support, you will never be able to properly secure it.

**Comment by TheZ3ro** [comment]
_Maybe the KeePassQuickUnlock author is reading this thread_ :stuck_out_tongue: 

In [May 2017](https://github.com/keepassxreboot/keepassxc/issues/488#issuecomment-299133413) I proposed to encrypt the masterkey with the pin.

In [November 2017](https://github.com/JanisEst/KeePassQuickUnlock/commit/2cd9c806e34d6e2995524c0b5296ff0702a7d82b) KeePassQuickUnlock starts encrypting the masterkey with the derived pin.


Anyway from what I can see the pin can't be used the first time you unlock a DB, am I right?
It keeps in memory the PIN-encrypted version of the CompositeKey/masterkey.
Maybe we can implement this as well, seems doable and relatively secure.

**Comment by mydexterid** [comment]
> Anyway from what I can see the pin can't be used the first time you unlock a DB, am I right?

Yes, at first you have to unlock with the normal method. After that you can use your quick unlock.

**Comment by Yorik-Code** [comment]
This feature is currently the only plugin that stops me from converting everything over to KeePassXC. Quick Unlock works really well for me as my 32 character master pwd is a little cumbersome to enter, but having a four character Quick Unlock is brilliant. I can see this making KeePassXC my goto for my Chromebook in Crouton.

The Keepass2Android implemenation is great as well, but I now also use the biometrics (fingerprint) ID for quick unlock as well, not saying that this is necessary here, but this little assistance increases usability for me a lot.

**Comment by Karreg** [comment]
Hello,

I'm testing KeePassXC, since I'm starting to see a lot of references to it (I'm a long run user of KeePass, and KeePass2Android). I'm coming here, because the quick unlock is definitely one of the two missing features versus keepass (+plugins) that could make me change software.

About the security, indeed quick unlock is less secure than full unlock.
BUT.
Full unlock can be so painful (with long passwords, just to input passwords that are less long) that probably a lot of peopledisable auto-lock of database. Which is even far less secure in the end.
In my case, I'm using a database every day at work, but it's too painful to full unlock it every few minutes (maximum value of inactivity being 9999s, so 2.5 hours, far from the 9 hours day of work).

About implementation, ideas here seem correct to me, also I'm using the two references given here (KeePassQuickUnlock and Keepass2Android) so I probably won't be of any help on this subject.

But about the quick unlock workflow, here is something that should reduce the attack possibilities:
1 - Quick unlock is not available if the database is closed
2 - On full unlock, some stuff is done and stored (from what I read, the latest option is encrypting master password with the PIN. Multiple encryption cycles could reinforce the security)
3 - On next unlock, since quick unlock data are available, the input is checked with the PIN
4 - If the input is valid, then database is unlocked with the decrypted master password
5 - If the input is wrong, quick unlock data are deleted from memory, and we're back to step 1 (number of retries may or may not be configurable. higher value means lower security)

It would also be great to have a setting to delete quick unlock data after a period of inactivity. For example, full unlock may be required every morning...

I'll be watching this issue, in the meantime I'll go back to KeePass with QuickUnlock and AutoTypeSearch (the other missing feature from KeePassXC for my usage)

Good work for the app by the way, it's less 90's than KeePass :)

**Comment by TheZ3ro** [comment]
> It would also be great to have a setting to delete quick unlock data after a period of inactivity. For example, full unlock may be required every morning...

I really like this proposal. Sadly implementing it is a little painful

**Comment by lourivellini** [comment]
Any progress on this?  Loving my keepass2android experience - wishing keepassXC would at least let me opt in.

**Comment by droidmonkey** [comment]
There are significant security concerns with implementing this because you essentially have to store the master key somewhere while the quick unlock is active. There are OS level means to do this, but its not trivial. 

We did implement TouchID quick unlock on macOS.

This will probably make it's way into the next major release after 2.5.0.

**Comment by MisterY** [comment]
As always, it's a convenience vs security. Adding an possibility for a user to decide is a fair option.
As previously mentioned, Windows now also has biometrics support. So, my real question is - should we track the request for the biometrics here or create a separate issue?
Thanks (for all the fish)!

**Comment by droidmonkey** [comment]
We have an issue for Windows Hello support already 😬

**Comment by MisterY** [comment]
I would like to reiterate an idea that @droidmonkey mentioned above. There is **no** need to store the master key anywhere, in memory or on disk. 
You should be able to encrypt the master password with a pin. Then store only the encrypted master password in the memory. When user uses the PIN to unlock, simply decrypt the stored value and then try to open the database with it. If it succeeds, great. The app has no knowledge of the PIN or the password. If it doesn't... well, too bad. :) 
It most probably means a wrong PIN and we're back at the beginning, needing to enter the full master password and encrypt with a PIN. 
I also see nothing wrong with having the user enter the PIN every time. This way it can be the same, or different every time.

@sudoforge mentioned that the database is *safer* by using the PIN. I disagree. It is much easier to catch a glance of a 3- or 4-digit PIN then a (relatively longer) pass phrase. Therefore, having the ability to enter a different PIN every time, would be more preferential to me.

**Comment by phoerious** [comment]
Actually, we do save the master key already, otherwise you couldn't save the database without entering your password every time.

**Comment by MisterY** [comment]
@phoerious, well, in that case, adding this feature will be risk-neutral. :)

**Comment by droidmonkey** [comment]
To clarify that master key is stored only when the database is unlocked. Once locked it gets obliterated. This is where quick unlock would instead take that stored master key, encrypt it with a preset pin, and store that for quick unlock usage.

**Comment by phoerious** [comment]
That is true, but quick unlock is still better than no lock at all and can be combined with a timeout after which you need to enter the full password again.

**Comment by electrofloat** [comment]
are you guys trying to reinvent the wheel?:
https://github.com/JanisEst/KeePassQuickUnlock

**Comment by sudoforge** [comment]
@MisterY said:

> @sudoforge mentioned that the database is safer by using the PIN.

I made no such statement. I have only stated that a PIN offers convenience.

**Comment by phoerious** [comment]
KeePass2 plugins are of no use to us.

**Comment by electrofloat** [comment]
@phoerious exactly. How about reading how the plugin works? Or reading the code itself?

**Comment by phoerious** [comment]
A quick-unlock feature is not a very complicated thing algorithmically, but one whose implementation is very specific to the host application. Reading the source code of the plug-in is not helpful, particularly since I happen to dislike the screenshots already.

**Comment by droidmonkey** [comment]
I read the code they are doing the exact same process I recommend

**Comment by MisterY** [comment]
@sudoforge, although my intention is not to go into this direction, you did write

>Windows 10 implements such a feature, where the user is only required to enter a small string of characters (a PIN in this particular case) that is - critically - not the same as their user password. This prevents keyloggers* or drive-by onlookers from being able to utilize the information they gather.

The reference is to the second sentence. To me that sounds like using a PIN helps prevent hacks although quite the opposite is the case, as I mentioned earlier. I believe PINs should be used only as a convenience in a relatively safe environment. Having written this, I will leave the thread to the directly related arguments, unless someone thinks that safety is also a concern here. 
I'm a strong believer in that the tools (a gun or a password manager) are not necessarily good or bad and all the responsibility lies on those using them. Provided there is enough information available.

**Comment by sudoforge** [comment]
@MisterY The last sentence in the portion of my comment that you quoted is accurate. A keylogger or on-looker that gains access to the PIN is not able to utilize the PIN outside of the current session. In the workflow that was being discussed when I left my earlier comments, the PIN would only be used as a secondary method of unlocking the database -- the first unlock of each session would still require the master passphrase.

**Comment by riedel** [comment]
Although yubikey has been mentioned across a few places: It would be great if Yubikey oder better FIDO2 (#3560  )without any additional PIN/Pass can be used as an alternative to vendor-specific mechanisms such Hello or TouchId for QuickUnlock .

**Comment by phoerious** [comment]
You can already use an HMAC-SHA1 YubiKey for the same purpose.

**Comment by MisterY** [comment]
> Although yubikey has been mentioned across a few places: It would be great if Yubikey oder better FIDO2 (#3560 )without any additional PIN/Pass can be used as an alternative to vendor-specific mechanisms such Hello or TouchId for QuickUnlock .

Hello and TouchId are not exactly "vendor-specific", they are part of the operating system. That would also imply that YubiKey is not vendor-specific?

>The YubiKey is a hardware authentication device manufactured by Yubico

https://en.wikipedia.org/wiki/YubiKey

**Comment by droidmonkey** [comment]
I looked into Windows Hello and it is not available in the Win32 API. We would have to have a C# companion app to access it. By vendor specific we mean OS vendor.

**Comment by droidmonkey** [comment]
This isn't something that we want to trivially implement. Ideally we would use the key stores available in each OS to store the encrypted master key. The implementation of touchid on macOS is a good example of how that would be achieved. Unfortunately, Windows has been moving all their API calls to UWP and Linux is a hot mess in general without clear standards.

Considering no one else has provided code to implement this I would consider this a "very difficult" feature.

**Comment by phoerious** [comment]
Microsoft moving all their new stuff to C# only is a huge problem for us in general, not only for Windows Hello.

**Comment by MisterY** [comment]
C# seems to work ok on Windows, Linux, and Android. Not sure about Mac. It's just another platform, right?

**Comment by phoerious** [comment]
KeePassXC is written in C++. We would need litter everything with .NET bindings or worst-case ditch Qt and rewrite our Windows GUI in C#. That's just not gonna happen. We don't want to maintain two UIs and one C# UI for all platforms would make KeePassXC just another KeePass with Mono as a dependency.

**Comment by MisterY** [comment]
True. And this is something where a modular architecture would help. 
As is visible from KeePass' example, support for plugins allows a wider range of interested individuals to create support for the features they badly need and alleviate the pressure from the core dev team.

**Comment by Karreg** [comment]
I agree. Especially when the position about security of quickunlock is strongly debatable...
The results of such discussions are that Keepass2 is still more pertinent on Linux than keepassxc, even with all the mono issues and incompatible plugins. Sad. :(

**Comment by phoerious** [comment]
> The results of such discussions are that Keepass2 and its plugins are still more pertinent on Linux than keepassxc,

Is that your personal opinion? Otherwise I'd like to see reliable data. Our position on plugins is non-negotioable and for good reason. We have seen the wild things that happen if you allow users to develop arbitrary plugins for a security application and I wouldn't let any single one of them onto my system. Besides, there are hardly any things you'd need a plugin for in KeePassXC, since it comes with most features already on board and batteries included.

**Comment by deathtrip** [comment]
> We have seen the wild things that happen if you allow users to develop arbitrary plugins for a security application and I wouldn't let any single one of them onto my system.
>
What do you mean exactly? Have there been any major problems with keepass2's plugins?

**Comment by Karreg** [comment]
Oh my... My answer was deleted. Oh well... Nevermind then.

**Comment by MisterY** [comment]
@Karreg , there was no email notification for any reply other than the ones I see here, so I would assume there was a problem before it got posted.

**Comment by phoerious** [comment]
> What do you mean exactly? Have there been any major problems with keepass2's plugins?

Yes, there have been. Questionable and dangerous security practices (like KeePassHttp), weird and inconsistent design and data format choices, and pretty much every plugin is abandonware.

> Oh my... My answer was deleted. Oh well... Nevermind then.

We did not delete any answers.

**Comment by Karreg** [comment]
True, in fact the editor was still open with the full version. So here is the initial content :)

---

Of course it's my personal opinion :)

I don't see any issue with your position, since it's your position on your own project... And your point about plugins is valid. But it does not mean that it's for the best.

However, what I see is some features that makes Keepass2 very powerful and/or improving the integration and usability of the tool being refused, or discussed for years, because discussion are never ending (autotype search, keyboard shortcuts, quick unlock, cloud connectivity...). And systems are evolving during this time, making any discussion obsolete even before implementation.

Let's face it, wanting a native application in C++ working everywhere is a good will, but the side effect is it's limiting its features to simplify the portability process. From an external point of view (mine, for what's it's worth to you), it just feels like KeepassX again.

That's why I find Keepass2 more pertinent on Linux than KeepassXC. More pain-in-the-ass to install, setup and maintain with the huge Mono stuff and distro workarounds, but also giving far more useful features, the one I gave before being the minimum to me in 2020. (Seriously, no cloud?). In fact, today, there's nothing keepassxc is doing that is not doable with keepass2 on linux.

I may seem a bit harsh to you and it's not my intention, it's just fursiously frustrating to see keepassxc being so far behind keepass2...

**Comment by phoerious** [comment]
> autotype search, keyboard shortcuts, quick unlock, cloud connectivity...

If you need these features and can't wait until we implement them, send a pull request. That's the same effort as writing a plugin, just better in every aspect whatsoever. Don't pin a lack of individual features on the unavailability of a plugin interface.

**Comment by Karreg** [comment]
I'm not saying it's due to the lack of plugin interface. As a user, I don't really care if it's a plugin or in the core component. I'm just pointing that these features are lacking. And By that, I'm not saying I miss them, because to be fair, I have abandonned KeepassXC just a few weeks after testing it. I'm just reacting to this discussion.

On the aspect of pull request, I thought we were over this. Asking people to learn a language just so they can implement something half-baked because they will still lack experience to write good code is just a way to say _we don't care really, we don't use them so we won't write them_. The result is just the same: features won't be implemented. It could be even worse: someone takes weeks or months to learn the language, write a new feature, and the PR could be refused because [insert any reason here]. 

This position was understandable when OSS softwares were done by developers for developers, and when languages were fewer. Today, It's not.

Also, taking years to implement something because it's OSS is fine. We all have a life, and OSS is developped on spare time, it's a huge implication towards community. I'm not discussing the delay. More the feeling that every feature is refused by default just because it's easier to say no, than to say yes but then have to deal with _when will it be done?_ every weeks. I understand that well, so it's really not a criticism of developpers, more of mindset...

Anyway, I have my needs covered, and KeepassXC is still a great tool, so I support you and your project. It just does not respond to my needs and my vision of what a password management tool should do... :)

My 2 cents.

**Comment by phoerious** [comment]
> More the feeling that every feature is refused by default just because it's easier to say no, than to say yes but then have to deal with when will it be done? every weeks. I understand that well, so it's really not a criticism of developpers, more of mindset...

Have you even been following the development of this project? I'm done discussing this.

**Comment by paddylandau** [comment]
I've tested KeePassXC, and I prefer it to KeePass2. However, I'm sticking to the latter specifically because of the lack of Quick Unlock. I know that previous commenters have said that having a Quick Unlock is insecure, but we have to bear in mind the typical user's requirements. I am almost a typical user; my father is a typical user. (A non-typical user would be welcome to avoid Quick Unlock.)

I find it tiresome to have to retype my (long!) passphrase every time I have to use a password, which is multiple times per day, sometimes over a couple of dozen times. To overcome this, I enable Quick Unlock on KeePass2. On KeePassXC, without Quick Unlock, I set the auto-lock to several hours. This is definitely less secure having a short auto-lock time combined with a Quick Unlock.

My father, an example of a typical user, would refuse to use a password manager without a Quick Unlock, as the typing would be too much for him.

Therefore, I am adding my +1 to this request. Until it is implemented, I unfortunately can't use KeePassXC and have to stick with KeePass2.

Thank you :)

**Comment by aamerabbas** [comment]
> I've tested KeePassXC, and I prefer it to KeePass2. However, I'm sticking to the latter specifically because of the lack of Quick Unlock. I know that previous commenters have said that having a Quick Unlock is insecure, but we have to bear in mind the typical user's requirements. I am almost a typical user; my father is a typical user. (A non-typical user would be welcome to avoid Quick Unlock.)
> 
> I find it tiresome to have to retype my (long!) passphrase every time I have to use a password, which is multiple times per day, sometimes over a couple of dozen times. To overcome this, I enable Quick Unlock on KeePass2. On KeePassXC, without Quick Unlock, I set the auto-lock to several hours. This is definitely less secure having a short auto-lock time combined with a Quick Unlock.
> 
> My father, an example of a typical user, would refuse to use a password manager without a Quick Unlock, as the typing would be too much for him.
> 
> Therefore, I am adding my +1 to this request. Until it is implemented, I unfortunately can't use KeePassXC and have to stick with KeePass2.
> 
> Thank you :)

+1 on this. Even as a power user, I could not imagine switching away from Bitwarden without this feature.

**Comment by droidmonkey** [comment]
I promise (say this a lot 😉) this will be the first thing I work on after 2.6 is released. We want this to be a used feature that is robust, not a bolt on. We did not make it into 2.6.0.

**Comment by jc11686** [comment]
For those that mentioned the security concerns, can the quick unlock feature incorporate using a Yubikey, biometric, or other additional means as an alternative quick unlock? 

I know that on my Windows PC I have Windows Hello unlock used for KeePass and it works great... but still there doesn't exist an ability like I'm suggesting - a means to use a Yubikey, or biometric, or a complex PIN, as a quick unlock method after the database has already been unlocked by the master password and then locked again due to the computer being locked.

My suggestion, if it can exist won't do away with the master password, but will allow other means to quick unlock the database (like a Yubikey), without setting the primary unlock method to require the Yubikey or biometric or PIN. First method will always be a master password + whatever else you've configured today (keyfile, Yubikey, other challenge-response method, etc). If the second unlock method after the session is locked could be one of these methods, and if they fail to work or the wrong PIN is used, then the quick unlock plugin / feature no longer allows quick unlock until the master password is used successfully again.

**Comment by paddylandau** [comment]
> I promise (say this a lot ) this will be the first thing I work on after 2.6 is released. We want this to be a used feature that is robust, not a bolt on. We did not make it into 2.6.0.

So… 2.6.2? (I ask hopefully.) Even if, in this short time frame, you can't incorporate a Yubikey or other key (as per the [previous comment](https://github.com/keepassxreboot/keepassxc/issues/488#issuecomment-709320490)), at least a three-character code that gives you only one chance to enter it correctly would be fantastic.

**Comment by phoerious** [comment]
2.7 the earliest. 2.6 is the current stable line and will not receive new features.

**Comment by lukefor** [comment]
It would be cool if this feature could integrate TPM, to give some degree of protection against hibernation attacks on unencrypted systems, and avoid writing an insecure database key to disk. Or as a simpler approach discard the quick unlock key on hibernate, but not on sleep?

**Comment by paddylandau** [comment]
> … discard the quick unlock key on hibernate, but not on sleep?

Discarding the Quick Unlock key on hibernation and sleep would _both_ need to be optional.

On Linux with encrypted storage (LUKS), hibernation leaves the entire system encrypted, so Quick Unlock isn't a security risk, and can remain enabled. In fact, in such a case, hibernation is significantly less a security risk than is sleep.

I would hope that the same applies to Windows with Bitlocker, and Mac with whatever it uses.

**Comment by lukefor** [comment]
> Discarding the Quick Unlock key on hibernation and sleep would _both_ need to be optional.
> 
> On Linux with encrypted storage (LUKS), hibernation leaves the entire system encrypted, so Quick Unlock isn't a security risk, and can remain enabled. In fact, in such a case, hibernation is significantly less a security risk than is sleep.
> 
> I would hope that the same applies to Windows with Bitlocker, and Mac with whatever it uses.

Agreed. But consideration should be made for unencrypted systems. People shouldn't need to use full disk encryption to safely enable a password manager feature

**Comment by R4ygen** [comment]
Also, apart from the big comfort bonus of not needing to type the long master password every time, a quick unlock feature is also a big security bonus for a specific situation that costantly fills my days: People watching you.

Since I've KeePass installed on my laptop that I use both while in class and at work, there are many occasions where a colleague or a classmate could peek the master password while typing it. And I don't feel like to be "that guy" that don't trust anyone and hides in a corner everytime I need a stored value, especially in the open space office at work.
Just have to full unlock it once and use a PIN for the rest of the day anytime I need it, makes me way less worried without sacrifice security (as long as I never leave my device unattended of course).

And maaybe, typing the master password less times even on your own pc, could be a very very minimal help aganist potential keyloggers, but yeah, just saying.

Anyway, I'll probably switch to KeePassXC from the original ASAP this will be implemented.

**Comment by phoerious** [comment]
Sleep and hibernate both cause KeePassXC to lock its database, so that wouldn't be an issue per se. However, sometimes the kernel hibernate before all applications processed the signal, which would only continue after waking up again.

**Comment by mlippert** [comment]
I use **keepass2android** mentioned in the original post, and I do like the Quick Unlock feature. I like that it's the last 3 characters of my long passphrase.
I also really like KeePassXC and have been using it (on linux) pretty much since it forked from KeePassX. I think you're doing a great job (thank you) and I'm very happy with the feature set that exists today. I've also got my mom using it on Windows. As a side note, I particularly like the TOTP functionality in KeePassXC.

I read through most of this discussion with interest and I look forward to seeing how you implement a quick unlock feature in 2.7. Having to set the quick unlock pin seems fine particularly if I will be able to set it to the last 3 characters of my passphrase if I so desire (just to keep consistency with my use of keepass2android)

**Comment by savchenko** [comment]
> if this feature could integrate TPM

That's an interesting idea actually.

@phoerious , @droidmonkey , do you think binding to a bunch of TCRs would be possible? This would allow to offload re-locking on power state changes and others to hardware.

**Comment by droidmonkey** [comment]
I'm going with a KISS approach

**Comment by W1sk3r** [comment]
I'm not sure how Linux handles secured strings, but I know that Windows (and Powershell) store secure strings that only that logged on user can read and decrypt. I know it isn't tied to the user's password because they change every N days and it still works... I've wondered how that can be done here if a PIN is set, or the last 3 of a pw, etc.

I know I don't know enough to speculate as far as how a Linux PC / workstation (non-LDAP-joined) would store the secure string for a user account. I need to add this to my ever growing "things to learn" list.

**Comment by onitake** [comment]
@W1sk3r Different platforms provide different kinds of native keychain services. On macOS, there is [Keychain Services](https://developer.apple.com/documentation/security/keychain_services). On Linux there are several, with [libsecret](https://developer.gnome.org/libsecret/) being the most widely supported - but there are others as well. And on Windows, there are several: [DPAPI](https://msdn.microsoft.com/en-us/library/ms995355.aspx), [Protected Storage](https://docs.microsoft.com/en-us/windows/win32/devnotes/pstore), [Credentials Management](https://docs.microsoft.com/en-us/windows/win32/secauthn/credentials-management).

There's a catch: If you already use KeypassXC as the native keychain via libsecret integration, then you obviously won't be able to use another keychain as the credential source _for_ KeypassXC. Kind of a chicken-egg problem, you know. :slightly_smiling_face:

**Comment by aamerabbas** [comment]
@W1sk3r, @onitake: Bitwarden is using keytar (https://github.com/atom/node-keytar) which provides an abstraction layer on top of the operating system's keychain.

**Comment by droidmonkey** [comment]
You can choose an easier password 🤔

**Comment by Corey-Keller** [comment]
> 
> 
> You can choose an easier password 🤔 or perhaps a yubikey

I like the idea of a YubiKey as an alternate option (as in the YubiKey would unlock a database that's secured by PW/key; basically I'm saying a YubiKey would make a good QuickUnlock option), but not as a primary due to the impracticality of always having it on me.

**Comment by Corey-Keller** [comment]
> 
> 
> Edit: why the down votes, the user sets their own password length. If you don't want to type a 100 character password, use a shorter password. 20 characters is more than sufficient with Argon2. Use a file key or yubikey for even more protection. Quick unlock still won't solve the initial unlock using an obscenely long password.

@droidmonkey 

1. QuickUnlock isn't supposed to solve the _initial_ unlock, but only subsequent unlocks of a DB that's been unlocked by the master password (I'm using "master password" here as a catchall for whatever combination of PW, Keyfile, and/or YubiKey you use)
2. I suspect the down votes were for the 'use a shorter password' sentiment. For the key to a password manager (and thus potentially access to **EVERYTHING** about you, digitally speaking) that should never **ever** be a 'solution' unless there is a _technical_ limitation based on length (like say things break if a password is over 65536 (2^16) characters or something)
3. I'm already using a keyfile in addition to the 100 char pw (which contains the entry to unlock the actual DB protected by a different keyfile and a ~2000 char pw). I just want an easier way to get back into my DBs _after_ I've got it unlocked once.

**Comment by droidmonkey** [comment]
I mean you do you, but you do realize that the actual key that encrypts your database is only 32 characters long? It is a transformed SHA-256 hash of your "key data". So if you want to be really technical about this, the maximum search space to unlock your database, no matter how long your password is or key file or yubikey, is 32 bytes. That is a huge amount of possibilities, by all accounts infinite, but your 100 character password and internal 1000 character password are not providing any additional protection beyond that search space.

**Comment by Corey-Keller** [comment]
> 
> 
> I mean you do you, but you do realize that the actual key that encrypts your database is only 32 bytes long? It is a transformed SHA-256 hash of your "key data". So if you want to be really technical about this, the maximum search space to unlock your database, no matter how long your password is or key file or yubikey, is 32 bytes. That is a huge amount of possibilities, by all accounts infinite, but your 100 character password and internal 2000 character password are not providing any additional protection beyond that search space.

I actually didn't know that. Neat! Thanks for the info. Correct me if I'm wrong, but that means that there's no security benefit to a master password (again, using that as a catchall for whatever combo of pw, keyfile etc. that make it up) with >256 bits of entropy correct?

But part of the desire for quick unlock is that we're dealing with 2 different attack surfaces:

1. a 'fully locked' database. This is the db as is when you first open KeePassXC (if using QuickUnlock for subsequent unlocks, if you're not then it applies to every unlock); here your attack surface is anyone who can get access to your database at all by any means, so high security is crucial. This is like hosting a server that is globally internet accessible; as long as you have the right credentials it doesn't matter where you are, you can get in.
2. A database that has been unlocked at least once on the machine, but is now locked. Here the attack surface is MUCH smaller (only those with access to the machine where it has been unlocked), so a somewhat less secure unlock method isn't as risky. This is like restricting access to a server to those in a specific place (server only accessible when on site or possibly connected to a secure VPN) or device. Instead of just anyone who has the credentials wherever they are being able to access the server, only those that have sufficiently proven themselves (on site, and thus approved by security, and/or connected from a company approved device) can access it.

**Comment by michaelk83** [comment]
> Correct me if I'm wrong, but that means that there's no security benefit to a master password with >256 bits of entropy correct?

This may be a bit outdated, but you can use [this chart](https://www.komando.com/wp-content/uploads/2021/03/Passwords-chart-1384x727.jpg) (or any number of similar ones) as a guide. Just as a couple examples, 15 lowercase letters should take 1000 years to crack, according to this chart, and 18 lowercase letters would take 23 *million* years to crack. That assumes the password is *completely random*.

15 lowercase letters is 1.7 * 10^21 combinations, which is equivalent to 70.5 bits of entropy. 18 lowercase letters is 3 * 10^25 combinations, or about 85 bits of entropy. So you get diminishing returns above 70 bits or so, and anything above 90 or so doesn't make much of a difference.

---

Anyway, I think [this comment](https://github.com/keepassxreboot/keepassxc/issues/488#issuecomment-617283762) summarizes the situation well. They're working on it (or at least are aware of the demand), but it's not an easy feature to add. The tricky part is handling the master password securely after the initial unlock.

If you can contribute a PR, you can ask the devs where to start. Otherwise, adding your own "I want this" comment isn't going to make this go any faster.

**Comment by michaelk83** [comment]
Update: It looks like fingerprint support will make this feature easier. MacOS TouchID [already functions as a sort of QuickUnlock](https://github.com/keepassxreboot/keepassxc/pull/1851). #6029 is trying to implement the same on Windows, and I suppose the [Linux side](https://github.com/keepassxreboot/keepassxc/issues/5991) will need to do something similar.

Once those is in place, regular QuickUnlock is pretty much the same thing, just without the biometrics. But there's a bit of a [TODO list](https://github.com/keepassxreboot/keepassxc/issues/3337#issuecomment-854550917) to finish them.

**Comment by MoneyAllDay** [comment]
With keepassxc-browser, would it be possible to also:

1. On browser, in a website with login input fields, click on the Keepass icon;
2. Manually input password to open database at first once;
3. Next time when filling user/pass fields just click on the icon and the extension would auto unlock the database and do its thing;
4. If the browser is closed the database password is "wiped from the cache". Requiring a start from step 1.

Of course the database password would be encrypted somehow in keepassxc-browser.

Having to type the password manually each time can get tedious, specially for those people who are on PC with timeouts enabled and 99% of the time opening the database is only related to the browser.

Any chance, this or a similar method is implemented in v2.7.0?

**Comment by aamerabbas** [comment]
>     3. Next time when filling user/pass fields just click on the icon and the extension would auto unlock the database and do its thing;

I will let the team comment on this, but I don't think this is possible. The extension just connects to the running instance of the app gets the credentials from there. This is part of the security as it ensures the extension itself doesn't actually hold any data. If the app is locked, the extension could not unlock it.

**Comment by 6677028** [comment]
I look forward to this feature. I hope the 2.7 update we will have. This is a very nice feature that Keepass and Bitwarden already have.

**Comment by christophschw** [comment]
As a workaround, I configured `fprintd` / `pam` for my root account and store the master key of the KDBX in root's GNOME Keyring. IMO, this is a good compromise between UX and security.

Editing my desktop file to this Exec line does the trick for initial unlocking via fingerprint:
`bash -c 'pkexec env DISPLAY=$DISPLAY XAUTHORITY=$XAUTHORITY secret-tool lookup application Keepass | keepassxc --pw-stdin ~/Nextcloud/Dokumente/Passwoerter.kdbx'`

![grafik](https://user-images.githubusercontent.com/13663130/181097910-5159b53b-c903-4634-95a9-1450bd115a0a.png)

Unfortunately, #2089 is blocking the UX after DB lock. Unlocking via this method fails with:
`Another instance of KeePassXC is already running.`

**Comment by phoerious** [comment]
Closing this, since KeePassXC has been supporting quick unlock on Windows and macOS since version 2.7.0 and we don't need this to track Linux support.

**Comment by paddylandau** [comment]
> Closing this, since KeePassXC has been supporting quick unlock on Windows and macOS since version 2.7.0 and we don't need this to track Linux support.

Where is LInux support tracked, please? Linux is already on version 2.7.1.

**Comment by michaelk83** [comment]
> Where is LInux support tracked, please?

#5991

**Comment by R4ygen** [comment]
> Closing this, since KeePassXC has been supporting quick unlock on Windows and macOS since version 2.7.0 and we don't need this to track Linux support.

Is a Pin unlock feature been implemented or does this refers just to the fingerprint / face unlock methods?
In the latter case, does there is another ticket open for a Pin unlock feature request?

**Comment by fnadde42** [comment]
Honestly, what is worse? That I use a PIN code that is not super safe since it requires storing master password in RAM..... OR... That I leave my DB completely open anyway, with it stored in RAM, since it's such a hassle to type in the master password all the time.

**Comment by michaelk83** [comment]
The current QuickUnlock uses Windows Hello on Windows, which AFAIK can be configured to use a PIN instead of a fingerprint. Don't know if that would actually work that way, but theoretically it should. (Can someone test this please?)

If the Linux implementation will be done via PolKit, it should be possible to configure it the same way, depending on your flavor of Linux and DE. In both cases, the preferred type of authentication is up to the host OS.

On macOS, QuickUnlock uses TouchID, which by the name of it is probably limited to biometrics (don't know). If so, then PIN unlock on macOS would require some additional code.

**Comment by R4ygen** [comment]
> Honestly, what is worse? That I use a PIN code that is not super safe since it requires storing master key in RAM..... OR... That I leave my DB completely open anyway, with the master password stored in RAM, since it's such a hassle to type in the master password all the time.

Afaik, the most used plugins on the OG KeePass work in a way that encrypts the hash of the master password with the PIN you provided. So yeah, it reduces the security of the database to a weaker password, but only locally and on an already booted system. And you only got 1 try with the pin or it relocks the db, but dunno if it helps in a direct memory attack.

Security notes for the [KeepassQuickUnlock plugin](https://github.com/JanisEst/KeePassQuickUnlock/issues/26#issuecomment-364114921), and the [LockAssist plugin](https://github.com/rookiestyle/lockassist/wiki/quick-unlock#security).



> The current QuickUnlock uses Windows Hello on Windows, which AFAIK can be configured to use a PIN instead of a fingerprint. Don't know if that would actually work that way, but theoretically it should. (Can someone test this please?)

Can confirm you can use a PIN for Windows Hello if you prefer it yes. My only worry is that it would be the same pin you use to unlock the pc, which is usually pretty weak and known to multiple users of that pc, and easily recoverable via "I've forgot my pin / password". It would be nice to have a separate pin only for unlocking Keepass, and maybe, isolated from the system, by being a part of Keepass itself instead of relying on OS dependent functions and security. (that would also be a big help for cross compatibility between systems, even for just the pin option)

**Comment by paddylandau** [comment]
> … the same pin you use to unlock the pc [is] known to multiple users of that pc

That in itself is a security issue beyond KeePassXC. Each user should have their own account with their own PIN, and KeePassXC cannot help if you share your PIN with others.

**Comment by amcsi** [comment]
> > …  the  same  pin  you  use  to  unlock  the  pc  [is]  known  to  multiple  users  of  that  pc
> 
> That in itself is a security issue beyond KeePassXC. Each user should have their own account with their own PIN, and KeePassXC cannot help if you share your PIN with others.

Irrelevant. An alternative PIN or short password should be an option as an additional layer of security.
